internal/optimization_test.go

```diff
@@ -0,0 +1,90 @@
+package internal
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDownloadOptimization(t *testing.T) {
+	// Setup test server
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("test content"))
+	}))
+	defer ts.Close()
+
+	tmpDir := t.TempDir()
+	lockDir := filepath.Join(tmpDir, "001")
+	err := os.MkdirAll(lockDir, 0755)
+	require.NoError(t, err)
+
+	testUrl := ts.URL + "/test.txt"
+	testContent := []byte("test content")
+	testFile := filepath.Join(tmpDir, "test.txt")
+	err = os.WriteFile(testFile, testContent, 0644)
+	require.NoError(t, err)
+
+	t.Run("Valid File Not Redownloaded", func(t *testing.T) {
+		lockPath := filepath.Join(lockDir, "grabit.lock")
+		lock, err := NewLock(lockPath, true)
+		require.NoError(t, err)
+
+		err = lock.AddResource([]string{testUrl}, RecommendedAlgo, nil, "test.txt")
+		require.NoError(t, err)
+
+		err = lock.Download(tmpDir, nil, nil, "")
+		require.NoError(t, err)
+	})
+
+	t.Run("Invalid File Redownloaded", func(t *testing.T) {
```

Comment:
I don't think we should blindly overwrite a file if it doesn't have the expected content. It's generally a bad idea to destroy content. In this case, grabit should error.

Reply:
> I don't think we should blindly overwrite a file if it doesn't have the expected content. It's generally a bad idea to destroy content. In this case, grabit should error.

Good clarification.  I think @777Denoiser did this because that's what I advised to do (if it doesn't validate, re-download it).  If we want grabit to error instead, cool.

Question: should grabit error on the first problem it finds?  Or should it still potentially download all the resources that it can, and list all errors that it found at the end of its run?